File: AVLTree/AVLTree.swift
class Node {
    var value: Int
    var left: Node?
    var right: Node?
    var height: Int

    init(value: Int) {
        self.value = value
        self.height = 1
    }
}

class AVLTree {
    func insert(_ node: Node?, _ value: Int) -> Node? {
        guard let node = node else {
            return Node(value: value)
        }

        if value < node.value {
            node.left = insert(node.left, value)
        } else if value > node.value {
            node.right = insert(node.right, value)
        }

        node.height = 1 + max(height(node.left), height(node.right))

        let balance = getBalance(node)

        if balance > 1 && value < node.left!.value {
            return rightRotate(node)
        }
        if balance < -1 && value > node.right!.value {
            return leftRotate(node)
        }
        if balance > 1 && value > node.left!.value {
            node.left = leftRotate(node.left)
            return rightRotate(node)
        }
        if balance < -1 && value < node.right!.value {
            node.right = rightRotate(node.right)
            return leftRotate(node)
        }

        return node
    }

    func inorder(_ node: Node?) {
        if let node = node {
            inorder(node.left)
            print(node.value, terminator: " ")
            inorder(node.right)
        }
    }

    func height(_ node: Node?) -> Int {
        return node?.height ?? 0
    }

    func getBalance(_ node: Node?) -> Int {
        return height(node?.left) - height(node?.right)
    }

    func rightRotate(_ y: Node) -> Node {
        let x = y.left
        let T2 = x?.right

        x?.right = y
        y.left = T2

        y.height = max(height(y.left), height(y.right)) + 1
        x?.height = max(height(x?.left), height(x?.right)) + 1

        return x!
    }

    func leftRotate(_ x: Node) -> Node {
        let y = x.right
        let T2 = y?.left

        y?.left = x
        x.right = T2

        x.height = max(height(x.left), height(x.right)) + 1
        y?.height = max(height(y?.left), height(y?.right)) + 1

        return y!
    }
}

// Usage
let tree = AVLTree()
var root: Node? = nil
let values = [10, 20, 30, 40, 50, 25]

for value in values {
    root = tree.insert(root, value)
}

tree.inorder(root)  // Output: 10 20 25 30 40 50

File: AVLTree/AVLTree.cpp
#include <iostream>
using namespace std;

class Node {
public:
    int value, height;
    Node* left;
    Node* right;

    Node(int val) {
        value = val;
        left = right = nullptr;
        height = 1;
    }
};

class AVLTree {
public:
    Node* insert(Node* node, int value) {
        if (node == nullptr)
            return new Node(value);

        if (value < node->value)
            node->left = insert(node->left, value);
        else if (value > node->value)
            node->right = insert(node->right, value);

        node->height = 1 + max(getHeight(node->left), getHeight(node->right));

        int balance = getBalance(node);

        if (balance > 1 && value < node->left->value)
            return rightRotate(node);

        if (balance < -1 && value > node->right->value)
            return leftRotate(node);

        if (balance > 1 && value > node->left->value) {
            node->left = leftRotate(node->left);
            return rightRotate(node);
        }

        if (balance < -1 && value < node->right->value) {
            node->right = rightRotate(node->right);
            return leftRotate(node);
        }

        return node;
    }

    void inorder(Node* root) {
        if (root != nullptr) {
            inorder(root->left);
            cout << root->value << " ";
            inorder(root->right);
        }
    }

private:
    int getHeight(Node* node) {
        if (node == nullptr)
            return 0;
        return node->height;
    }

    int getBalance(Node* node) {
        if (node == nullptr)
            return 0;
        return getHeight(node->left) - getHeight(node->right);
    }

    Node* rightRotate(Node* y) {
        Node* x = y->left;
        Node* T2 = x->right;

        x->right = y;
        y->left = T2;

        y->height = max(getHeight(y->left), getHeight(y->right)) + 1;
        x->height = max(getHeight(x->left), getHeight(x->right)) + 1;

        return x;
    }

    Node* leftRotate(Node* x) {
        Node* y = x->right;
        Node* T2 = y->left;

        y->left = x;
        x->right = T2;

        x->height = max(getHeight(x->left), getHeight(x->right)) + 1;
        y->height = max(getHeight(y->left), getHeight(y->right)) + 1;

        return y;
    }
};

int main() {
    AVLTree tree;
    Node* root = nullptr;
    int values[] = {10, 20, 30, 40, 50, 25};

    for (int value : values) {
        root = tree.insert(root, value);
    }

    tree.inorder(root);  // Output: 10 20 25 30 40 50
    return 0;
}